File: Cocoa/AppFrameworks/UIKit/TableView/Cell/YJUITableViewHeaderFooterView.m
//
//  YJUITableViewHeaderFooterView.m
//  YJCocoa
//
//  HomePage:https://github.com/937447974/YJCocoa
//  YJ技术支持群:557445088
//
//  Created by 阳君 on 2018/7/18.
//  Copyright © 2018年 YJCocoa. All rights reserved.
//

#import "YJUITableViewHeaderFooterView.h"
#import "YJUITableViewManager.h"
#import "YJNSFoundationOther.h"

@implementation UITableViewHeaderFooterView (YJUITableView)

+ (CGFloat)tableViewManager:(YJUITableViewManager *)tableViewManager heightForCellObject:(YJUITableCellObject *)cellObject {
    switch (cellObject.createCell) {
        case YJUITableViewCellCreateClass:
            return tableViewManager.tableView.sectionHeaderHeight;
        case YJUITableViewCellCreateXib: {
            NSArray<UITableView *> *array = [[NSBundle mainBundle] loadNibNamed:cellObject.cellName owner:nil options:nil];
            return CGRectGetHeight(array.firstObject.frame);
        }
        case YJUITableViewCellCreateSoryboard: {
            UITableViewCell *cell = [tableViewManager.tableView dequeueReusableCellWithIdentifier:cellObject.cellName];
            return CGRectGetHeight(cell.frame);
        }
    }
}

@end

@implementation YJUITableViewHeaderFooterView

#pragma mark - getter & setter
- (NSString *)reuseIdentifier {
    NSString *reuseIdentifier = [super reuseIdentifier];
    if (reuseIdentifier) return reuseIdentifier;
    return YJNSStringFromClass(self.class);
}

#pragma mark - YJUITableView
+ (YJUITableViewCellCreate)cellCreate {
    return YJUITableViewCellCreateClass;
}

+ (CGFloat)tableViewManager:(YJUITableViewManager *)tableViewManager heightForCellObject:(YJUITableCellObject *)cellObject {
    if ([self.class isEqual:YJUITableViewHeaderFooterView.class]) {
        return 0;
    }
    return [super tableViewManager:tableViewManager heightForCellObject:cellObject];
}

- (void)reloadDataWithCellObject:(YJUITableCellObject *)cellObject tableViewManager:(YJUITableViewManager *)tableViewManager {
    _cellObject = cellObject;
    _tableViewManager = tableViewManager;
}

@end

File: YJCocoa/Classes/AppFrameworks/UIKit/TableView/YJUITableViewHeaderFooterView.swift
//
//  YJUITableViewHeaderFooterView.swift
//  YJCocoa
//
//  HomePage:https://github.com/937447974/YJCocoa
//  YJ技术支持群:557445088
//
//  Created by 阳君 on 2019/5/23.
//  Copyright © 2016-现在 YJCocoa. All rights reserved.
//

import UIKit

@objc extension UITableViewHeaderFooterView {
    
    /// 获取 YJUITableCellObject
    public class func cellObject() -> YJUITableCellObject {
        return YJUITableCellObject(cellClass: self)
    }
    
    /// 获取YJUITableCellObject并自动填充模型
    public class func cellObject(withCellModel cellModel:AnyObject) -> YJUITableCellObject {
        let co = self.cellObject()
        co.cellModel = cellModel
        return co
    }
    
    /// 获取 header 的显示高
    open class func tableViewManager(_ tableViewManager: YJUITableViewManager, heightWith cellObject: YJUITableCellObject) -> CGFloat {
        return tableViewManager.tableView.sectionHeaderHeight
    }
    
    /// 刷新 UITableViewHeaderFooterView
    @objc func tableViewManager(_ tableViewManager: YJUITableViewManager, reloadWith cellObject: YJUITableCellObject) {}
    
}

open class YJUITableViewHeaderFooterView: UITableViewHeaderFooterView {
    
    private(set) var cellObject: YJUITableCellObject!
    private(set) var tableViewManager: YJUITableViewManager!
    
    open override var reuseIdentifier: String? {
        return super.reuseIdentifier ?? NSStringFromClass(self.classForCoder)
    }
    
    open override func tableViewManager(_ tableViewManager: YJUITableViewManager, reloadWith cellObject: YJUITableCellObject) {
        super.tableViewManager(tableViewManager, reloadWith: cellObject)
        self.tableViewManager = tableViewManager
        self.cellObject = cellObject
    }
    
}
